File: solutions/python/005-rle/solution.py
"""Run-length encoding / decoding."""

def encode(s: str) -> str:
    if not s:
        return ""
    result: list[str] = []
    i = 0
    while i < len(s):
        ch = s[i]
        j = i + 1
        while j < len(s) and s[j] == ch:
            j += 1
        result.append(ch + str(j - i))
        i = j
    return "".join(result)


def decode(s: str) -> str:
    if not s:
        return ""
    result: list[str] = []
    i = 0
    while i < len(s):
        ch = s[i]
        if not ch.isalnum():
            raise ValueError("expected alphanumeric character")
        i += 1
        if i >= len(s) or not s[i].isdigit():
            raise ValueError("missing count after character")
        count_str = ""
        while i < len(s) and s[i].isdigit():
            count_str += s[i]
            i += 1
        n = int(count_str)
        if n < 1:
            raise ValueError("count must be at least 1")
        result.append(ch * n)
    return "".join(result)

File: solutions/javascript/005-rle/solution.mjs
function isAlnum(c) {
  return /[A-Za-z0-9]/.test(c);
}

export function encode(s) {
  if (!s) return "";
  let result = "";
  let i = 0;
  while (i < s.length) {
    const ch = s[i];
    let j = i + 1;
    while (j < s.length && s[j] === ch) j++;
    result += ch + String(j - i);
    i = j;
  }
  return result;
}

export function decode(s) {
  if (!s) return "";
  let result = "";
  let i = 0;
  while (i < s.length) {
    const ch = s[i];
    if (!isAlnum(ch)) throw new Error("expected alphanumeric character");
    i++;
    if (i >= s.length || !/\d/.test(s[i])) throw new Error("missing count after character");
    let countStr = "";
    while (i < s.length && /\d/.test(s[i])) {
      countStr += s[i];
      i++;
    }
    const n = Number(countStr);
    if (n < 1) throw new Error("count must be at least 1");
    result += ch.repeat(n);
  }
  return result;
}
